File: api/src/anime_api/proxy.py
"""HLS proxy: playlists, segments and subtitles flow through the API.

The upstream video CDN 403s without the embed site's Referer and sends no
CORS headers, so every stream URL is routed through ``/stream/<enc>?ref=``.
Playlists are rewritten in-flight so nested URLs keep flowing through the
proxy; segments/subtitles are piped with Range support and Referer fallback
(the CDN may demand any of several embed origins).
"""

from __future__ import annotations

import re
from typing import AsyncIterator
from urllib.parse import quote, unquote, urljoin

import httpx
from fastapi import Request
from fastapi.responses import JSONResponse, Response, StreamingResponse

from .http import USER_AGENT, Http
from .resolver import KNOWN_EMBED_ORIGINS

CHUNK = 64 * 1024


def decode_target(path: str) -> str | None:
    """Extract the upstream URL from ``/stream/<urlencoded>``."""
    raw = unquote(path.lstrip("/"))
    if not raw:
        return None
    return raw if raw.startswith(("http://", "https://")) else None


def referer_for(target: str, ref_query: str | None) -> str:
    if ref_query and ref_query.startswith("http"):
        return ref_query
    return target.split("//", 1)[0] + "//" + target.split("//", 1)[1].split("/", 1)[0] + "/"


def rewrite_m3u8(body: str, base_url: str, referer: str) -> str:
    """Rewrite every URI in an m3u8 so nested requests use /stream/ too."""
    ref_qs = f"?ref={quote(referer, safe='')}"

    def wrap(url: str) -> str:
        absolute = urljoin(base_url, url)
        return f"/stream/{quote(absolute, safe='')}{ref_qs}"

    out_lines: list[str] = []
    for line in body.splitlines():
        stripped = line.strip()
        if not stripped:
            out_lines.append(line)
            continue
        if stripped.startswith("#"):
            # rewrite URI="..." attributes (EXT-X-KEY, EXT-X-MAP, ...)
            rewritten = re_uri(stripped, wrap)
            out_lines.append(rewritten)
        else:
            out_lines.append(wrap(stripped))
    return "\n".join(out_lines) + "\n"


def re_uri(line: str, wrap) -> str:
    return re.sub(r'URI="([^"]+)"', lambda m: f'URI="{wrap(m.group(1))}"', line)


async def proxy_stream(
    http: Http, request: Request, path: str, ref: str | None
) -> Response:
    """Handle one /stream request (playlist rewrite or segment pipe)."""
    target = decode_target(path)
    if target is None:
        return JSONResponse({"error": "invalid stream path"}, status_code=400)

    referer = referer_for(target, ref)
    is_playlist = ".m3u8" in target.split("?", 1)[0]
    is_vtt = ".vtt" in target.split("?", 1)[0]

    if is_playlist:
        try:
            body = await http.text(target, referer=referer, use_cache=False)
        except Exception as exc:
            return JSONResponse(
                {"error": f"proxy failure: {exc}"}, status_code=502
            )
        return Response(
            content=rewrite_m3u8(body, target, referer),
            media_type="application/vnd.apple.mpegurl",
            headers={"Cache-Control": "no-store"},
        )

    # segments + subtitles: pipe through with Range support and Referer fallback
    client: httpx.AsyncClient = http._client
    range_header = request.headers.get("range")
    referers = [referer]
    target_origin = referer_for(target, None)
    if target_origin not in referers:
        referers.append(target_origin)
    for extra in KNOWN_EMBED_ORIGINS:
        if extra not in referers:
            referers.append(extra)

    headers = {"User-Agent": USER_AGENT, "Accept": "*/*"}
    if range_header:
        headers["Range"] = range_header

    last_error: Exception | None = None
    for ref_try in referers:
        headers["Referer"] = ref_try
        headers["Origin"] = ref_try.rstrip("/")
        try:
            upstream = await client.send(
                client.build_request("GET", target, headers=headers), stream=True
            )
            if upstream.status_code in (401, 403, 404):
                await upstream.aclose()
                last_error = Exception(f"upstream {upstream.status_code}")
                continue
            upstream.raise_for_status()
        except httpx.HTTPStatusError as exc:
            last_error = exc
            continue
        except httpx.HTTPError as exc:
            last_error = exc
            continue

        # stream it through
        resp_headers = {
            k: v
            for k, v in upstream.headers.items()
            if k.lower()
            in (
                "content-type",
                "content-length",
                "content-range",
                "accept-ranges",
                "etag",
                "last-modified",
            )
        }

        async def iter_bytes() -> AsyncIterator[bytes]:
            try:
                async for chunk in upstream.aiter_bytes(CHUNK):
                    yield chunk
            finally:
                await upstream.aclose()

        return StreamingResponse(
            iter_bytes(),
            status_code=upstream.status_code,
            headers=resp_headers,
            media_type=upstream.headers.get("content-type") or "application/octet-stream",
        )

    status = 403 if "upstream 403" in str(last_error) else 502
    return JSONResponse({"error": f"proxy failure: {last_error}"}, status_code=status)


# image proxy: cover hosts send no CORS headers and sometimes check Referer
IMAGE_HOST_ALLOWLIST = (
    "hianime.at",
    "hianime.to",
    "anipixcdn.co",
    "s4.anilist.co",
    "cdn.myanimelist.net",
    "img.anili.st",
    "media.kitsu.app",
    "wsrv.nl",
)


def image_allowed(url: str) -> bool:
    try:
        host = url.split("//", 1)[1].split("/", 1)[0].lower()
    except IndexError:
        return False
    return any(host == h or host.endswith("." + h) for h in IMAGE_HOST_ALLOWLIST)


async def proxy_image(http: Http, path: str) -> Response:
    """Handle one /img request with a host allowlist."""
    target = decode_target(path)
    if target is None or not image_allowed(target):
        return JSONResponse({"error": "invalid image path"}, status_code=400)

    try:
        upstream = await http._client.get(
            target,
            headers={"User-Agent": USER_AGENT, "Accept": "image/*", "Referer": "https://hianime.at/"},
        )
        upstream.raise_for_status()
    except Exception as exc:
        return JSONResponse({"error": f"image fetch failed: {exc}"}, status_code=502)

    return Response(
        content=upstream.content,
        media_type=upstream.headers.get("content-type") or "image/jpeg",
        headers={"Cache-Control": "public, max-age=86400"},
    )

File: site/server/proxy.ts
/**
 * HLS proxy: streams upstream m3u8 playlists and segments through the server,
 * rewriting nested URLs so they keep flowing through the proxy. The upstream
 * video CDN 403s without the embed site's Referer, so callers pass it as
 * ?ref= and we propagate it upstream (falling back to the target origin and
 * then to known embed origins on 403).
 */
import type { Request, Response } from "express";
import { runtimeConfig } from "./config.js";

const UA = runtimeConfig.userAgent;

function upstreamUrl(req: Request): string | null {
  const raw = req.path.replace(/^\/+/, "");
  if (!raw) return null;
  let decoded: string;
  try {
    decoded = decodeURIComponent(raw);
  } catch {
    return null;
  }
  if (!/^https?:\/\//i.test(decoded)) return null;
  return decoded;
}

function refererFor(target: string, req: Request): string {
  const ref = req.query.ref;
  if (typeof ref === "string" && /^https?:\/\//i.test(ref)) return ref;
  return new URL(target).origin + "/";
}

/** Rewrite URLs inside an m3u8 so nested requests also go through /stream/. */
export function rewriteM3u8(body: string, baseUrl: string, referer: string): string {
  const refQs = `?ref=${encodeURIComponent(referer)}`;
  return body
    .split(/\r?\n/)
    .map((line) => {
      const trimmed = line.trim();
      if (!trimmed) return line;
      if (trimmed.startsWith("#")) {
        // rewrite URI="..." attributes inside tags (e.g. EXT-X-KEY, EXT-X-MAP)
        return line.replace(/URI="([^"]+)"/g, (_full, uri: string) => {
          const abs = new URL(uri, baseUrl).toString();
          return `URI="/stream/${encodeURIComponent(abs)}${refQs}"`;
        });
      }
      const abs = new URL(trimmed, baseUrl).toString();
      return `/stream/${encodeURIComponent(abs)}${refQs}`;
    })
    .join("\n");
}

async function pumpUpstream(
  res: Response,
  target: string,
  referer: string,
  range: string | undefined,
  signal: AbortSignal,
): Promise<void> {
  const upstream = await fetch(target, {
    headers: {
      "User-Agent": UA,
      Referer: referer,
      Origin: new URL(referer).origin,
      ...(range ? { Range: range } : {}),
    },
    signal,
    redirect: "follow",
  });
  if (!upstream.ok || upstream.body === null) {
    throw new Error(`upstream ${upstream.status}`);
  }
  const contentType = upstream.headers.get("content-type") ?? "";
  if (contentType) res.setHeader("Content-Type", contentType);
  const contentLength = upstream.headers.get("content-length");
  const contentRange = upstream.headers.get("content-range");
  if (contentLength) res.setHeader("Content-Length", contentLength);
  if (contentRange) res.setHeader("Content-Range", contentRange);
  res.status(upstream.status === 206 ? 206 : 200);

  // Respect TCP backpressure so fast upstreams don't build a giant queue,
  // and stop pulling bytes as soon as the client disconnects (seek/close)
  // so the freed bandwidth goes to the segments actually being watched.
  const reader = upstream.body.getReader();
  try {
    for (;;) {
      const { done, value } = await reader.read();
      if (done || res.closed || res.destroyed) break;
      if (res.write(Buffer.from(value))) continue;
      await new Promise<void>((resolve) => {
        res.once("drain", resolve);
        res.once("close", resolve);
      });
    }
  } finally {
    reader.cancel().catch(() => undefined);
  }
  if (!res.closed && !res.destroyed) res.end();
}

export async function streamHandler(req: Request, res: Response): Promise<void> {
  const target = upstreamUrl(req);
  if (!target) {
    res.status(400).json({ error: "invalid stream path" });
    return;
  }

  const targetPath = new URL(target).pathname;
  const isM3u8 = targetPath.endsWith(".m3u8");
  const isVtt = targetPath.endsWith(".vtt");
  const referer = refererFor(target, req);

  const ctrl = new AbortController();
  const timer = setTimeout(() => ctrl.abort(), isM3u8 || isVtt ? 15_000 : 60_000);

  try {
    if (isM3u8) {
      const upstream = await fetch(target, {
        headers: { "User-Agent": UA, Referer: referer },
        signal: ctrl.signal,
        redirect: "follow",
      });
      if (!upstream.ok) throw new Error(`upstream ${upstream.status}`);
      const text = await upstream.text();
      res.setHeader("Content-Type", upstream.headers.get("content-type") ?? "application/vnd.apple.mpegurl");
      res.setHeader("Cache-Control", "no-store");
      res.send(rewriteM3u8(text, target, referer));
      return;
    }

    // segments + subtitle files: passthrough (vtt needs no URL rewriting)
    const referers = [referer];
    const targetOrigin = new URL(target).origin + "/";
    if (!referers.includes(targetOrigin)) referers.push(targetOrigin);
    // known embed origins the CDN may require
    for (const extra of ["https://zokoanime.video/", "https://megaplay.buzz/"]) {
      if (!referers.includes(extra)) referers.push(extra);
    }

    let lastErr: Error | null = null;
    for (const ref of referers) {
      try {
        await pumpUpstream(res, target, ref, req.headers.range, ctrl.signal);
        return;
      } catch (err) {
        lastErr = err as Error;
        if (res.headersSent) break; // too late to retry
      }
    }
    throw lastErr ?? new Error("upstream failed");
  } catch (err) {
    if (!res.headersSent) {
      const status = /upstream 403/.test((err as Error).message) ? 403 : 502;
      res.status(status).json({ error: `proxy failure: ${(err as Error).message}` });
    } else {
      res.end();
    }
  } finally {
    clearTimeout(timer);
  }
}
